### ISSUE TITLE
Port to Hyper's tokio branch

### ISSUE THREAD

**Issue by untitaker** [iron/iron#501]
Rough plan:

* Add a new trait `AsyncHandler` with a method `handle_async`. There is a `impl<H: Handler> AsyncHandler for H` that just delegates an invocation of `Handler::handle(self, ...)` to a threadpool. (where is the threadpool stored?). The idea is that users can still impl synchronous handlers using `Handler`, but drop down to `AsyncHandler` if they really care about performance.

  * Important: Implement `AsyncHandler` for functions, etc
* All middlewares become async. Synchronous middlewares are not supported. `AroundMiddleware` takes `Box<AsyncHandler>` and returns `Box<AsyncHandler>`
* Everything else in Iron goes async as well.
* The response object will have to change somehow, probably. Hopefully response modifiers can abstract the changes away.

**Comment by Hoverbear** [comment]
This will be a big job and introduce significant breaking changes.

**Comment by ernestas-poskus** [comment]
If you plan to make middlewares async it might brake expected behavior where authorization or alike middleware must go first.

**Comment by untitaker** [comment]
Middlewares would still be applied in order.

**Comment by Hoverbear** [comment]
Right. By async we mean the middlewares themselves should support being async, so they should not block, for example. They would still be done in order.

**Comment by hayd** [comment]
Hyper's tokio branch is now in master. https://github.com/hyperium/hyper/commit/2d2d5574a698e74e5102d39b9a9ab750860d92d1

**Comment by untitaker** [comment]
#523 updates to async Hyper, for now just using a threadpool

**Comment by MJDSys** [comment]
I've started digging into the meat of this issue (Async Iron on Async Hyper, based off of #523).  However, I'm running some lifetime issues.

Basically, when my ```async_handle``` function returns, I can't have references to either the handler or the request hanging around, as the response isn't constructed until later.  For requests, I've tried stepping around this by moving in a ```Request``` instead of passing a ```&mut Request```, and producing a ```Request``` as part of the output.  I'm not super happy about this though, as it leaks the details of the ```Chain``` into ```async_handle```'s interface.  Also there is now an ugly tuple to return.  The trait is currently:
```rust
/// `AsyncHandler`s are responsible for handling all requests by returning a future that yields the actual response.
pub trait AsyncHandler: Send + 'static {
    /// Produce a `Future` that handles the given request that produces an appropriate response or error.
    fn async_handle(&self, Request) -> BoxIronFuture<(Request, Response)>;
}

// where BoxIronFuture is a helper type defined as:
pub type BoxIronFuture<T> = Box<futures::Future<Item=T, Error=IronError>>;
```

However, if anything in ```&self``` is needed, it needs to be cloned into the future.  This is a problem implementing Chain, as the ```Vec```s of middleware are needed to decide how to proceed, but I can't keep a reference to them.  I've currently fixed it by avoiding ```BoxFuture,``` which requires the ```Future``` to be ```Send``` (since we are not currently sending the future across threads, which I'm not sure is a valid assumption?).  I then store a ```Vec<Rc<*Middleware>>``` instead of ```Vec<Box<*Middleware>>```, but I'm not super happy about that.  If I use ```BoxFuture, Vec<Arc<*Middleware>>``` isn't enough, as ```*Middleware``` isn't ```Sync```.  I'm not sure if it's a problem to require ```*Middleware``` to be ```Sync``` though.

I tried making the ```&self``` in ```async_handle``` ```'static```, but due to Hyper not allowing its handler to be ```'static```, that doesn't work.  So it seems I'm stuck with an ```Rc```.

I'm thinking the ```Vec<Rc<*>>``` needs to be a ```Rc<Vec<Box<*>>>``` to avoid some ```Box<Future<>>``` in ```Chain```, but that's an optimization with a separate discussion.

Also reem/rust-plugin#14 remains a problem.  For now, I've removed the ```Extensible``` implementation and used a ```SendMap``` directly.

**Comment by bfrog** [comment]
Will this let me build my own listener per thread now? Its been a long time pain for me that everything using iron must be Sync+Send if shared among handlers when what I'd really like is per thread state rather than global state.

**Comment by MJDSys** [comment]
@bfrog My current implementation in #527 currently is single threaded, and the AsyncHandler's don't require Sync+Send.  However, the Request does need to be Send due to needing Handler implementations needing to be run on a thread pool, and thus requiring the Request to be sendable across threads (and thus anything you put in the request's extensions needs to be Sendable).

I don't know if the Iron devs will want to change that.

**Comment by bfrog** [comment]
Thats perfect. No I don't extend the Request's themselves, I'd like to have per thread handlers that look something like.... struct MyHandler { db_conn: Conn } where conn doesn't need to be shared among a bunch of threads. Your changes make it sound like this will work perfectly with that.

**Comment by MJDSys** [comment]
Yep, that should work fine.  The only problem is that you can't do any long work in the async_handle function, as you block the IO as well.  I'm looking to start using my branch for a project, and I'm looking to use tokio_postgres to talk to my database for that reason.  You might also have to share that one connection across several requests, but at least you don't have to worry about threading :)
